contributors/design-proposals/auth/bound-service-account-tokens.md

```diff
@@ -0,0 +1,193 @@
+core apiserver. Once this API exists, a client of the apiserver will request an
+attenuated token for it's own use. The API will enforce required attenuations,
+e.g. audience and time binding.
+
+## Token attenuations
+
+### Audience binding
+
+Tokens issued from this API will be audience bound. Audience of requested tokens
+will be bound by the `aud` claim. The `aud` claim is an array of strings
+(usually URLs) that correspond to the intended audience of the token. A
+recipient of a token is responsible for verifying that it identifies as one of
+the values in the audience claim, and should otherwise reject the token. The
+TokenReview API will support this validation.
+
+### Time binding
+
+Tokens issued from this API may be time bound. Time validity of these tokens
+will be claimed in the following fields:
+
+* `exp`: expiration time
+* `nbf`: not before
+* `iat`: issued at
+
+A recipient of a token should verify that the token is valid at the time that
+the token is presented, and should otherwise reject the token. The TokenReview
+API will support this validation.
+
+Cluster administrators will be able to configure:
+
+* Whether or not non-expiring tokens are should be issued
+* The maximum validity duration for expiring tokens
+
+### Object binding
+
+Tokens issued from this API may be bound to a Kubernetes object in the same
+namespace as the service account. The name, group, version, kind and uid of the
+object will be embedded as claims in the issued token. A token bound to an
+object will only be valid for as long as that object exists.
+
+Only a subset of object kinds will support object binding. Initially the only
+kinds that will be supported are:
+
+* v1/Pod
+* v1/Secret
+
+The TokenRequest API will validate this binding.
+
+### Key binding
+
+Tokens issued from this API will OPTIONALLY (per the behest of the requestor)
+be key bound. The `cnf` claim will carry information that a recipient of a
+token might need to validate that the token presenter possesses a specific key.
+A recipient may require accepted tokens are key bound.
+
+## API Changes
+
+### Add `tokenrequests.authentication.k8s.io`
+
+We will add an imperative API (a la TokenReview) to the
+`authentication.k8s.io` API group:
+
+```golang
+type TokenRequest struct {
+  Spec   TokenRequestSpec
+  Status TokenRequestStatus
+}
+
+type TokenRequestSpec struct {
+  // Audiences are the intendend audiences of the token. A token issued
+  // for multiple audiences may be used to authenticate against any of
+  // the audiences listed. This implies a high degree of trust between
+  // the target audiences.
+  Audience []string
+
+  // ValidityDuration is the requested duration of validity of the request. The
+  // token issuer may return a token with a different validity duration so a
+  // client needs to check the 'expiration' field in a response.
+  ValidityDuration metav1.Duration
+
+  // BoundObject is a reference to an object that the token will be bound to.
```

Comment:
Couldn't a token be bounded to multiple objects? A token given to a node for a pod running a service account could be tied to:

* The node
* The pod
* The service account

Reply:
A token should always be bound to a service account i.e. the service account needs to exist for the service account token to be valid. I consider these bindings to be extras (e.g. pod, node, secret). I don't think N bindings over 1 binding adds extra value to warrant the complexity. Deleting a node today should cause all pods on the node to be deleted almost instantaneously. Is there any other usecases for multiple bound objects?

Reply:
>  I don't think N bindings over 1 binding adds extra value to warrant the complexity.

I tend to agree (and also don't want unbounded items to resolve at validation time)